File: modules/fbs-core/api/src/main/scala/de/thm/ii/fbs/services/misc/UserCleanupService.scala
package de.thm.ii.fbs.services.v2.misc

import de.thm.ii.fbs.services.persistence.UserService
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.{Autowired, Value}
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service

import java.sql.Date
import java.time.Instant

@Service
class UserCleanupService {
  @Value("${privacy.cleanUsers.enabled}")
  private val enabled: Boolean = false
  @Value("${privacy.cleanUsers.secondsAfterLastLogin}")
  private val secondsAfterLastLogin: Int = 15768000

  @Autowired
  private val userService: UserService = null

  private val logger = LoggerFactory.getLogger(this.getClass)

  @Scheduled(fixedDelayString = "PT1D") // Runs every day
  def cleanOldUsers(): Unit = {
    val now = new Date(Instant.now().minusSeconds(secondsAfterLastLogin).getEpochSecond*1000)
    val usersToDelete = userService.getUsersWithLastLoginBefore(now)
    logger.info(s"Found ${usersToDelete.length} users to delete.")
    if (usersToDelete.nonEmpty) {
      usersToDelete.foreach(u => userService.delete(u.id))
      logger.info(s"Deleted ${usersToDelete.map(u => u.username).mkString(", ")}")
    }
  }
}

File: modules/fbs-core/api/src/main/kotlin/de/thm/ii/fbs/services/v2/misc/UserCleanupService.kt
package de.thm.ii.fbs.services.v2.misc

import de.thm.ii.fbs.services.v2.persistence.UserRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime

@Service
open class UserCleanupService(
    private val userRepository: UserRepository,
    @Value("\${security.userCleanup.enabled:false}")
    private val enabled: Boolean,
    @Value("\${security.userCleanup.inactiveDaysLimit:365}")
    private val inactiveDaysLimit: Long
) {
    private val logger = LoggerFactory.getLogger(UserCleanupService::class.java)

    @Scheduled(cron = "\${security.userCleanup.checkSchedule:0 0 3 * * SUN}")
    @Transactional
    open fun cleanupInactiveUsers() {
        if (!enabled) {
            logger.debug("Benutzer-Cleanup ist deaktiviert.")
            return
        }

        val cutoffDate = LocalDateTime.now().minusDays(inactiveDaysLimit)
        logger.info("Starte Anonymisierung für Benutzer inaktiv seit $cutoffDate.")

        try {
            val inactiveUsers = userRepository.findUsersToAnonymize(cutoffDate)
            val affectedRows = userRepository.anonymizeInactiveUsers(cutoffDate)
            logger.info("Cleanup erfolgreich: $affectedRows von ${inactiveUsers.size} inaktiven Benutzern anonymisiert.")
        } catch (e: Exception) {
            logger.error("Fehler beim Cleanup: ${e.message}", e)
        }
    }
}